feat: More admin options for organizations

hariombalhara opened the pull request:
## What does this PR do?
Stacked over https://github.com/calcom/cal.com/pull/12358. Please merge that first.
### Admin Options for Organizations
- Ability to publish an org 
- Edit
	- Name
	- orgAutoAcceptEmail
	- Slug
		-  Delete previous domain and record on Vercel and Cloudflare
		-  Create new domain and record on Vercel and Cloudflare
- Ability to delete an organization.
	- Delete domain on Vercel
	- Delete DNS record on cloudflare

[Loom Demo](https://www.loom.com/share/3af28427fb7b41ed944551316166e42b)

### Other fixes
- Fixes booker URLs shown in Other Team Profile View and Create New Team Views
- Disable Error Redaction locally.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

hariombalhara commented:
Current dependencies on/for this PR:
* main
  * **PR #12358** <a href="https://app.graphite.dev/github/pr/calcom/cal.com/12358?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #12424** <a href="https://app.graphite.dev/github/pr/calcom/cal.com/12424?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/calcom/cal.com/12424?utm_source=stack-comment).

hariombalhara commented on packages/lib/redactError.ts:
```diff
@@ -18,7 +20,7 @@ export const redactError = <T extends Error | unknown>(error: T) => {
   }
   log.debug("Type of Error: ", error.constructor);
-  if (shouldRedact(error)) {
+  if (shouldRedact(error) && IS_PRODUCTION) {
```
Let's redact on production only. Keeping things convenient locally.

hariombalhara commented on packages/trpc/server/routers/viewer/organizations/_router.tsx:
```diff
@@ -114,7 +109,7 @@ export const viewerOrganizationsRouter = router({
     return handler(opts);
   }),
-  listOtherTeams: authedOrgAdminProcedure.query(async (opts) => {
+  listOtherTeams: authedProcedure.query(async (opts) => {
```
It was unnecessarily causing 403 for all non admin members. Instead, let the route itself handle it and return empty array for non admin members,

hariombalhara commented on packages/trpc/server/routers/viewer/organizations/_router.tsx:
```diff
@@ -64,14 +67,6 @@ export const viewerOrganizationsRouter = router({
-    const handler = await importHandler(namespaced("adminVerify"), () => import("./adminVerify.handler"));
-    return handler(opts);
-  }),
   listMembers: authedProcedure.input(ZListMembersSchema).query(async (opts) => {
```
Moved below to keep all admin endpoints together.

hariombalhara commented on packages/trpc/server/routers/viewer/teams/delete.handler.ts:
```diff
@@ -73,11 +38,7 @@ export const deleteHandler = async ({ ctx, input }: DeleteOptions) => {
 
   const deletedTeamMetadata = teamMetadataSchema.parse(deletedTeam.metadata);
 
-  if (IS_PRODUCTION)
```
Let us allow testing domain deletion locally as well.

hariombalhara commented on packages/features/ee/organizations/pages/settings/admin/AdminOrgEditPage.tsx:
```diff
@@ -0,0 +1,107 @@
+import type { Team } from "@prisma/client";
```
Borrowed some of the stuff from AdminOrgPage

hariombalhara commented on packages/features/ee/organizations/pages/settings/admin/AdminOrgPage.tsx:
```diff
@@ -54,7 +92,7 @@ function AdminOrgTable() {
                   <br />
                   <span className="text-muted">
-                    {org.slug}.{extractDomainFromWebsiteUrl}
+                    {org.slug}.{subdomainSuffix()}
```
Fixes the domain shown

hariombalhara commented on packages/trpc/server/routers/viewer/organizations/adminVerify.schema.ts:
```diff
@@ -1,10 +1,7 @@
-
 export const ZAdminVerifyInput = z.object({
   orgId: z.number(),
-  status: statusSchema,
```
Wasn't being used.

PeerRich commented:
great work, super pumped about it

sean-brydon commented on packages/trpc/server/routers/viewer/organizations/adminDelete.handler.ts:
```diff
@@ -0,0 +1,70 @@
+        id: user.id,
+      },
+      data: {
+        username: `${user.username || ""}-${user.id}`,
```
I think we should actually specific this -> You can very easily get a cal account with no username set and having a user with `-519` isnt something we normally allow i dont think 

```suggestion
        username: `${user.username || "nameless_user"}-${user.id}`,
```

hariombalhara replied:
What's the case when there is no username ? Does it happen when we invite a user?

I actually wanted to avoid `nameless` but I think it is a better approach when we know for sure that having null username is possible.

sean-brydon replied:
Its very rare but you can replicate it by
1) invite a new user
2) new user doesnt want to join a team 
3) resets password
4) logs in 

We should probably work on re-directing to onboarding if no username is set. Or ensure user.username exists bfore we delete them? Not sure the best approach around that

sean-brydon commented on packages/trpc/server/routers/viewer/organizations/adminGet.handler.ts:
```diff
@@ -0,0 +1,60 @@
+  const org = await prisma.team.findUnique({
+    where: {
+      id: input.id,
+      AND: [
```
Just a little idea i had when reading this - i wonder if it would be easier to query by  just team ID -> parse metadata -> check its set and throw from there. 

(Probably ms saved no biggy)

hariombalhara replied:
Yeah, makes a lot of sense actually.

hariombalhara replied:
Fixed

hariombalhara commented on packages/lib/domainManager/deploymentServices/cloudflare.ts:
Yeah I thought about using classes but decided against it initially as it would add some complexity. IMHO, when we add another service we could refactor then.

hariombalhara commented on packages/trpc/server/routers/viewer/organizations/adminDelete.handler.ts:
```diff
@@ -0,0 +1,70 @@
+
+export default adminDeleteHandler;
+
+async function renameUsersToAvoidUsernameConflicts(users: { id: number; username: string | null }[]) {
```
I doubt that it would happen because X is user.id which is always incrementing even when users are deleted.

hariombalhara replied:
Added an inline comment for this as well.

Udit-takkar commented on packages/features/ee/organizations/pages/settings/admin/AdminOrgPage.tsx:
```diff
@@ -145,3 +228,54 @@ const AdminOrgList = () => {
+              Teams that are member of this organization will also be deleted along with their event-types
+            </li>
+            <li>
+              Users that were part of the organization will not be deleted and their event-types will also
+              remain intact.
```
NIT: use i18n

hariombalhara replied:
It already is. Notice `i18nKey`

hariombalhara committed:
Add more features in org admin

ThyMinimalDev commented on packages/lib/domainManager/deploymentServices/vercel.ts:
```diff
@@ -0,0 +1,147 @@
+      statusCode: 400,
+    });
+  }
+  if (error.code === "domain_already_in_use") {
+    return true;
+  }
```
should we have extra information here on the reason for this error.code if statement to just return true, maybe some comments or logs

ThyMinimalDev commented on packages/lib/domainManager/deploymentServices/vercel.ts:
```diff
@@ -0,0 +1,147 @@
+    throw new Error("Missing env var: PROJECT_ID_VERCEL");
+  }
+
+  if (!env.TEAM_ID_VERCEL) {
+    // TEAM_ID_VERCEL is optional
+  }
```
are we planing to add something in this if statement ?

ThyMinimalDev commented on packages/trpc/server/routers/viewer/organizations/adminDelete.handler.ts:
```diff
@@ -0,0 +1,80 @@
+
+export default adminDeleteHandler;
+
+async function renameUsersToAvoidUsernameConflicts(users: { id: number; username: string | null }[]) {
```
no need to have any custom check for premium username here I suppose ?

hariombalhara replied:
Yeah I don't think so. I don't think premium usernames are anyways applicable for an organization.

ThyMinimalDev commented on packages/trpc/server/routers/viewer/organizations/adminUpdate.handler.ts:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  const foundOrg = organizationsWithSameSlug[0];
+  if (!foundOrg) {
+    return;
+  }
```
shouldn't we return an error here ? or some information to display on the frontend ?

hariombalhara replied:
This is actually the happy case as there is no org that has the same slug. So, no conflict and no need to throw an error.

hariombalhara committed:
Pr feedback addressed

hariombalhara commented:
Thanks for the review @ThyMinimalDev. Addressed all comments.

Also, FYI it's a stacked PR. So, I would really appreciate if you could [review the target branch PR ](https://github.com/calcom/cal.com/pull/12358) as well.

hariombalhara commented:
@ThyMinimalDev No longer a stacked PR now. It's targeted to main now. If you could do a final review that would be great.

ThyMinimalDev reviewed (approved):
looking good
